src/asmcup/genetics/Gene.java

```diff
@@ -10,16 +12,27 @@ public Gene(byte[] dna, int gen, float score) {
 		this.gen = gen;
 		this.score = score;
 	}
-	
+
+	@Override
 	public int compareTo(Gene other) {
 		float d = score - other.score;
 		
-		if (d == 0) {
+		if (Math.abs(d) < 0.00000001f) {
 			return 0;
 		} else if (d < 0) {
 			return 1;
 		}
 		
 		return -1;
 	}
+
+	@Override
+	public boolean equals(Object obj) {
+		return obj instanceof Gene && compareTo((Gene) obj) == 0;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(score);

```

Comment:
There will be *tons* of genes with score zero in the GA. If anybody decides to do use any construct that assumes hashes to almost always be different, that construct will be bogged down by hash collision handling.